{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "446f0991",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Welcome. Use this program to design oligos for making single sgRNA plasmids. \n",
    "# DO NOT use this to make oligos for pooled sgRNA  libraries. \n",
    "# Those primers are slightly different\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cc0eac3a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/willboyd/Documents/Lab/Jupyter notebooks/Pioneer Factor sgRNA Library'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "cwd = os.getcwd()\n",
    "cwd\n",
    "#this is just to help you find where your excel files are\n",
    "#you will need to download the Calabrese library if you are designing CRISPRa sgRNA and\n",
    "#the Dolcetto library if you are designing CRISPRi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "401cdfe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#For the new user, there are only 2 parts you need to change for each run\n",
    "#the genes you want to target and the CRISPRi/a library you are taking the sgRNA sequence from\n",
    "\n",
    "#Type the names of the Genes you are interested in targeting here\n",
    "\n",
    "Target_Genes = ['HNF4A', 'FOXA2', 'CDX2', 'GATA6'] \n",
    "\n",
    "#Load the CRISPRi/a library you are taking the sgRNA sequence from here\n",
    "\n",
    "import pandas as pd\n",
    "df = pd.read_excel(\"/Users/willboyd/Documents/Lab/Jupyter notebooks/Pioneer Factor sgRNA Library/Calabrese Library A.xlsx\")\n",
    "\n",
    "# Load data from Genomewide sgRNA Library file\n",
    "#Make sure that the sgRNA are coming from the Calabrese Library if you are looking for CRISPRa sgRNA\n",
    "# and from the Dolcetto if you are looking for CRISPRi sgRNA\n",
    "\n",
    "\n",
    "#if this part works you should be able to run the rest with no issues\n",
    "#also make sure to restart the kernel for every run, im not good at coding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "676ff707",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Convert the 'Annotated Gene Symbol' column to strings\n",
    "df['Annotated Gene Symbol'] = df['Annotated Gene Symbol'].astype(str)\n",
    "\n",
    "# Sort the DataFrame by Annotated Gene Symbol\n",
    "df = df.sort_values(by=\"Annotated Gene Symbol\").reset_index(drop=True)\n",
    "\n",
    "# Dictionary to keep track of occurrences of each Annotated Gene Symbol\n",
    "gene_count = {}\n",
    "\n",
    "# List to store the modified Annotated Gene Symbols\n",
    "modified_gene_symbols = []\n",
    "\n",
    "# Iterate through each row to modify Annotated Gene Symbols\n",
    "for gene_symbol in df['Annotated Gene Symbol']:\n",
    "    if gene_symbol not in gene_count:\n",
    "        gene_count[gene_symbol] = 1\n",
    "    else:\n",
    "        gene_count[gene_symbol] += 1\n",
    "    \n",
    "    # Append the suffix to the Annotated Gene Symbol\n",
    "    modified_gene_symbols.append(f\"{gene_symbol}-{gene_count[gene_symbol]}\")\n",
    "\n",
    "# Add the modified Annotated Gene Symbols to the DataFrame\n",
    "df['Modified Annotated Gene Symbol'] = modified_gene_symbols\n",
    "\n",
    "# Select relevant columns and rename for clarity\n",
    "organized_df = df[['Modified Annotated Gene Symbol', 'Barcode Sequence', 'Annotated Gene Symbol']]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "78d38d97",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Create a filter that matches any of the words in the 'Annotated Gene Symbol' column\n",
    "filter_mask = organized_df['Annotated Gene Symbol'].isin(Target_Genes)\n",
    "\n",
    "# Filter the DataFrame\n",
    "filtered_df = df[filter_mask]\n",
    "\n",
    "filtered_df =filtered_df[['Modified Annotated Gene Symbol', 'Annotated Gene Symbol', 'Barcode Sequence']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "afdd55ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Modified Annotated Gene Symbol</th>\n",
       "      <th>Annotated Gene Symbol</th>\n",
       "      <th>Barcode Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CDX2-1-fwd</td>\n",
       "      <td>CDX2</td>\n",
       "      <td>CACCGAATGCAAATTATGTTTCGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CDX2-1-rev</td>\n",
       "      <td>CDX2</td>\n",
       "      <td>AAACCTCGAAACATAATTTGCATTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDX2-2-fwd</td>\n",
       "      <td>CDX2</td>\n",
       "      <td>CACCGAGGGGAGGGAGGCTCGTGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CDX2-2-rev</td>\n",
       "      <td>CDX2</td>\n",
       "      <td>AAACCTCACGAGCCTCCCTCCCCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CDX2-3-fwd</td>\n",
       "      <td>CDX2</td>\n",
       "      <td>CACCGGGGGAGGGAGGCTCGTGAGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDX2-3-rev</td>\n",
       "      <td>CDX2</td>\n",
       "      <td>AAACTCTCACGAGCCTCCCTCCCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>FOXA2-1-fwd</td>\n",
       "      <td>FOXA2</td>\n",
       "      <td>CACCGAGAGGGAGCGCGAGAGAGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>FOXA2-1-rev</td>\n",
       "      <td>FOXA2</td>\n",
       "      <td>AAACCCCTCTCTCGCGCTCCCTCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>FOXA2-2-fwd</td>\n",
       "      <td>FOXA2</td>\n",
       "      <td>CACCGCTCTCGCGCTCCCTCTCCCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>FOXA2-2-rev</td>\n",
       "      <td>FOXA2</td>\n",
       "      <td>AAACAGGGAGAGGGAGCGCGAGAGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FOXA2-3-fwd</td>\n",
       "      <td>FOXA2</td>\n",
       "      <td>CACCGGAGGGAGCGCGAGAGAGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>FOXA2-3-rev</td>\n",
       "      <td>FOXA2</td>\n",
       "      <td>AAACTCCCTCTCTCGCGCTCCCTCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GATA6-1-fwd</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>CACCGGTGAGTCCAATCAGGAGCCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GATA6-1-rev</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>AAACGGGCTCCTGATTGGACTCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GATA6-2-fwd</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>CACCGAAGGGTGCCAGGCTGTGGGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GATA6-2-rev</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>AAACACCCACAGCCTGGCACCCTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GATA6-3-fwd</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>CACCGAACAAACAGCGCTCGCCGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GATA6-3-rev</td>\n",
       "      <td>GATA6</td>\n",
       "      <td>AAACTTCGGCGAGCGCTGTTTGTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>HNF4A-1-fwd</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>CACCGGCCCAGCCTATCCACCGGCG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>HNF4A-1-rev</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>AAACCGCCGGTGGATAGGCTGGGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>HNF4A-2-fwd</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>CACCGCCGCCGGTGGATAGGCTGGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>HNF4A-2-rev</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>AAACCCCAGCCTATCCACCGGCGGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>HNF4A-3-fwd</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>CACCGCCGCCCAGCCTATCCACCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>HNF4A-3-rev</td>\n",
       "      <td>HNF4A</td>\n",
       "      <td>AAACCCGGTGGATAGGCTGGGCGGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Modified Annotated Gene Symbol Annotated Gene Symbol  \\\n",
       "0                      CDX2-1-fwd                  CDX2   \n",
       "1                      CDX2-1-rev                  CDX2   \n",
       "2                      CDX2-2-fwd                  CDX2   \n",
       "3                      CDX2-2-rev                  CDX2   \n",
       "4                      CDX2-3-fwd                  CDX2   \n",
       "5                      CDX2-3-rev                  CDX2   \n",
       "6                     FOXA2-1-fwd                 FOXA2   \n",
       "7                     FOXA2-1-rev                 FOXA2   \n",
       "8                     FOXA2-2-fwd                 FOXA2   \n",
       "9                     FOXA2-2-rev                 FOXA2   \n",
       "10                    FOXA2-3-fwd                 FOXA2   \n",
       "11                    FOXA2-3-rev                 FOXA2   \n",
       "12                    GATA6-1-fwd                 GATA6   \n",
       "13                    GATA6-1-rev                 GATA6   \n",
       "14                    GATA6-2-fwd                 GATA6   \n",
       "15                    GATA6-2-rev                 GATA6   \n",
       "16                    GATA6-3-fwd                 GATA6   \n",
       "17                    GATA6-3-rev                 GATA6   \n",
       "18                    HNF4A-1-fwd                 HNF4A   \n",
       "19                    HNF4A-1-rev                 HNF4A   \n",
       "20                    HNF4A-2-fwd                 HNF4A   \n",
       "21                    HNF4A-2-rev                 HNF4A   \n",
       "22                    HNF4A-3-fwd                 HNF4A   \n",
       "23                    HNF4A-3-rev                 HNF4A   \n",
       "\n",
       "             Barcode Sequence  \n",
       "0   CACCGAATGCAAATTATGTTTCGAG  \n",
       "1   AAACCTCGAAACATAATTTGCATTC  \n",
       "2   CACCGAGGGGAGGGAGGCTCGTGAG  \n",
       "3   AAACCTCACGAGCCTCCCTCCCCTC  \n",
       "4   CACCGGGGGAGGGAGGCTCGTGAGA  \n",
       "5   AAACTCTCACGAGCCTCCCTCCCCC  \n",
       "6   CACCGAGAGGGAGCGCGAGAGAGGG  \n",
       "7   AAACCCCTCTCTCGCGCTCCCTCTC  \n",
       "8   CACCGCTCTCGCGCTCCCTCTCCCT  \n",
       "9   AAACAGGGAGAGGGAGCGCGAGAGC  \n",
       "10  CACCGGAGGGAGCGCGAGAGAGGGA  \n",
       "11  AAACTCCCTCTCTCGCGCTCCCTCC  \n",
       "12  CACCGGTGAGTCCAATCAGGAGCCC  \n",
       "13  AAACGGGCTCCTGATTGGACTCACC  \n",
       "14  CACCGAAGGGTGCCAGGCTGTGGGT  \n",
       "15  AAACACCCACAGCCTGGCACCCTTC  \n",
       "16  CACCGAACAAACAGCGCTCGCCGAA  \n",
       "17  AAACTTCGGCGAGCGCTGTTTGTTC  \n",
       "18  CACCGGCCCAGCCTATCCACCGGCG  \n",
       "19  AAACCGCCGGTGGATAGGCTGGGCC  \n",
       "20  CACCGCCGCCGGTGGATAGGCTGGG  \n",
       "21  AAACCCCAGCCTATCCACCGGCGGC  \n",
       "22  CACCGCCGCCCAGCCTATCCACCGG  \n",
       "23  AAACCCGGTGGATAGGCTGGGCGGC  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Copy each row\n",
    "copied_fdf = filtered_df.copy()\n",
    "# Append \"-fwd\" to the original Modified Annotated Gene Symbol and \"rev\" to the copied Modified Annotated Gene Symbol\n",
    "copied_fdf['Modified Annotated Gene Symbol'] = copied_fdf['Modified Annotated Gene Symbol'] + \"-rev\"\n",
    "filtered_df['Modified Annotated Gene Symbol'] = filtered_df['Modified Annotated Gene Symbol'] + \"-fwd\"\n",
    "\n",
    "#Make the rev barcode sequence the reverse complement of the fwd\n",
    "complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "copied_fdf['Barcode Sequence'] = copied_fdf['Barcode Sequence'].apply(lambda x: ''.join(complement[base] for base in x[::-1]))\n",
    "\n",
    "#append a CACCG to the front of the fwd oligo to add the sticky hands and the U6 start codon \"G\"\n",
    "#append a AAAC to the front and C to the back of the rev oligo to add the sticky hands and start codon\n",
    "filtered_df['Barcode Sequence'] = 'CACCG' + filtered_df['Barcode Sequence'].astype(str)\n",
    "copied_fdf['Barcode Sequence'] = 'AAAC' + copied_fdf['Barcode Sequence'].astype(str) + \"C\"\n",
    "\n",
    "#Combine, organize and export the finalized oligo sequences\n",
    "final_df = pd.concat([filtered_df, copied_fdf], ignore_index=True)\n",
    "final_df = final_df[['Modified Annotated Gene Symbol', 'Annotated Gene Symbol', 'Barcode Sequence']]\n",
    "final_df = final_df.sort_values(by=\"Modified Annotated Gene Symbol\").reset_index(drop=True)\n",
    "final_df.to_excel(\"primer_data.xlsx\", index=False)\n",
    "\n",
    "#check the product here\n",
    "final_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09ac920b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2556c63",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3f8c7f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
